src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

```diff
@@ -1309,22 +1322,172 @@ void WholeBodyDynamicsDevice::readContactPoints()
     // robot is divided by the F/T sensors.
 
     measuredContactLocations.clear();
-
-    // For now just put the default contact points
+    int numberOfContacts=0;
     size_t nrOfSubModels = estimator.submodels().getNrOfSubModels();
 
-    for(size_t subModel = 0; subModel < nrOfSubModels; subModel++)
+    // read skin
+    iCub::skinDynLib::skinContactList *scl =this->portContactsInput.read(false); //scl=null could also mean no new message
+    if(scl && useSkinContact)
     {
-        bool ok = measuredContactLocations.addNewContactInFrame(estimator.model(),
-                                                                subModelIndex2DefaultContact[subModel],
-                                                               iDynTree::UnknownWrenchContact(iDynTree::FULL_WRENCH,iDynTree::Position::Zero()));
-        if( !ok )
+
+        //< \todo TODO check for envelope?
+        last_reading_skin_contact_list_Stamp = yarp::os::Time::now();
+        if(scl->empty())   // if no skin contacts => leave the old contacts but reset pressure and contact list
+        {
+
+
+            if (contactsReadFromSkin.empty())
+            {
+
+                for(size_t subModel = 0; subModel < nrOfSubModels; subModel++)
+                {
+                    bool ok = measuredContactLocations.addNewContactInFrame(estimator.model(),
+                                                                            subModelIndex2DefaultContact[subModel], //frameIndex in iDynTree
+                                                                            iDynTree::UnknownWrenchContact(iDynTree::FULL_WRENCH,iDynTree::Position::Zero()));
+                    if( !ok )
+                    {
+                        yWarning() << "wholeBodyDynamics: Failing in adding default contact for submodel " << subModel;
+                    }
+
+
+
+
+                }
+                return;
+            }
+
+            //< \todo TODO this (using the last contacts if no contacts are detected) should be at subtree level, not at global level??
+            // ask what is intended with this TODO
+            else
+            {
+                for(iCub::skinDynLib::skinContactList::iterator it=contactsReadFromSkin.begin(); it!=contactsReadFromSkin.end(); it++)
+                {
+                    it->setPressure(0.0);
+                    it->setActiveTaxels(0);
+                    numberOfContacts++;
+                }
+                yInfo() << "wholeBodyDynamics: numberOfContacts in contactsReadFromSkin from previous contacts = "<<numberOfContacts;
```

Comment:
I will leave this part commented since it will probably come useful when we try again with Joans stuff 

```
/*else
        {
             yDebug() << "wholeBodyDynamics: number of contacts in submodel "<<subModel<<" = "<<contacts_for_given_subModel[subModel];
        }*/
```